### ISSUE TITLE
Context provided in _app.js can't be consumed in pages in SSR

### ISSUE THREAD

**Issue by jaydenseric** [zeit/next.js#4194]
React v16.3 context provided in `_app.js` can be consumed and rendered in pages on the client, but is undefined in SSR. This causes React SSR markup mismatch errors.

Note that context can be universally provided/consumed *within* `_app.js`, the issue is specifically when providing context in `_app.js` and consuming it in a page such as `pages/index.js`.

- [x] I have searched the [issues](https://github.com/zeit/next.js/issues?q=is%3Aissue) of this repository and believe that this is not a duplicate.

## Expected Behavior
Context provided in _app.js should be consumable in pages both on the server for SSR and when browser rendering.

## Current Behavior
Context provided in _app.js is undefined when consumed in pages for SSR. It can only be consumed for client rendering.

## Steps to Reproduce (for bugs)

In `_app.js`:

```jsx
import App, { Container } from 'next/app'
import React from 'react'
import TestContext from '../context'

export default class MyApp extends App {
  render () {
    const { Component, pageProps } = this.props
    return (
      <Container>
        <TestContext.Provider value="Test value.">
          <Component {...pageProps} />
        </TestContext.Provider>
      </Container>
    )
  }
}
```

In `context.js`:

```jsx
import React from 'react'

export default React.createContext()
```

In `pages/index.js`:

```jsx
import TestContext from '../context'

export default () => (
  <TestContext.Consumer>
    {value => value}
  </TestContext.Consumer>
)
```

Will result in:

<img width="570" alt="screen shot 2018-04-23 at 2 29 29 pm" src="https://user-images.githubusercontent.com/1754873/39107043-d45bd2a6-4702-11e8-85a8-f04ef11bcb4b.png">

## Context

A large motivation for the `_app.js` feature is to be able to provide context persistently available across pages. It's unfortunate the current implementation does not support this basic use case.

I'm attempting to isomorphically provide the cookie in context so that [`graphql-react`](https://github.com/jaydenseric/graphql-react) `<Query />` components can get the user's access token to make GraphQL API requests. This approach used to work with separately decorated pages.

## Your Environment

| Tech    | Version |
|---------|---------|
| next    | v6.0.0-canary.5 |
| node    | v9.11.1 |

**Comment by timneutkens** [comment]
Is it because _app.js provides its own context? 🤔

**Comment by timneutkens** [comment]
That actually can't be the problem when you nest inside `<Container>` though 🤔

**Comment by jaydenseric** [comment]
@timneutkens as the bundling seems to work for the client, does that mean we could split the server bundles in a similar way to the client?

**Comment by timneutkens** [comment]
Unfortunately it doesn't work as expected with commonschunk on the server. I have some ideas to fix this once and for all, but haven't had time to implement it yet.

**Comment by tz5514** [comment]
Will this be fixed? This problem break lots of behavior from module's local variable. 
ex: react-act https://github.com/pauldijou/redux-act#types

**Comment by RustyDev** [comment]
Maybe this will help. Seems like this project is down in there... [https://github.com/faceyspacey/react-universal-component](https://github.com/faceyspacey/react-universal-component)

**Comment by brad-decker** [comment]
@timneutkens i'm not super familiar but i'm willing to do a deep dive to help figure out a solution. Is there any pointers / additional details you could share to help me get started, i'm gonna get started researching the issue on my own but any knowledge transfer would be helpful.

**Comment by timneutkens** [comment]
So, since Next 5 we run webpack 2 times, once for the server, once for the client. 
Both have different configurations. 

On the server webpack will pack every page into it's own bundle (`entry`point), which ends up in `.next/dist`. This bundle holds everything, including all imported components etc.

Say you have 2 pages that both import `Header` it will end up in both page 1 and page 2's bundle.

Because of this module state / singletons don't work, since if you server render page 1, and it holds a module state inside the component, for example:

```js
let timesRendered = 0

export default () => {
  timesRendered++
  return <div>test</div>
}
```

It won't shared between server rendering of 2 pages.

From what I've understood looking at the new context API, this is exactly what the context API depends on (local module state). So if you have `_app.js` and you initialize it there, and then try to access the context from `pages/index.js`, they'll have 2 different context module states.

I hope this explains the issue.

I'm not entirely sure how we can fix this issue. In Next 4 and below we emitted every possible file into `.next/dist`, and then, because we only applied Babel, the paths when `import`ing were still correct, but with webpack this works a little different, since it parses import/export and turns it into a dependency tree, and also bundles to a file, not 1-1 into the specified directory.

**Comment by brad-decker** [comment]
@timneutkens thanks for this writeup. i'm digging in now and might have additional questions for you as I work through it.

**Comment by joeporpeglia** [comment]
Is it possible to create a similar ["main" commons chunk](https://github.com/zeit/next.js/blob/canary/server/build/webpack.js#L266-L295) for the server bundles? This issue doesn't seem like a problem for the client bundles (`.next/bundles/pages/*.js` and `.next/static/commons/main-*.js`) since the `createContext()` call only exists in the main chunk. I haven't been able to get a working example, but I'll keep messing around with the idea.

**Comment by brad-decker** [comment]
@joeporpeglia @timneutkens -> I used dynamic to import the context provider in the app and the consumer in the page to test how that modified the bundle. As expected the code related to the consumer initialization was put into its own chunk in .next/dist but the problem isn't solved. I gave my context consumer a obnoxious name to help with searching the dist files.

```js
import React, { Component, createContext } from 'react';

const BradContext = createContext({ x: 3 });

export class BradProvider extends Component {
  state = {
    x: 5,
  };
  render() {
    return (
      <BradContext.Provider value={this.state}>
        {this.props.children}
      </BradContext.Provider>
    )
  }
}

export const BradConsumer = BradContext.Consumer;
```

The state is kept entirely within this file. 

In _app.js
```js
import App, { Container } from 'next/app';
import dynamic from 'next/dynamic';

const BradProvider = dynamic(import('../components/context').then(r => r.BradProvider));

export default class MyApp extends App {
  render() {
    const { Component, pageProps } = this.props;
    return (
      <Container>
        <BradProvider>
          <Component {...pageProps} />
        </BradProvider>
      </Container>
    )
  }
}
```

in index.js:
```js
import React, { Component } from 'react';
import dynamic from 'next/dynamic';

const BradConsumer = dynamic(import('../components/context').then(r => r.BradConsumer));

export default class Test extends Component {
  render() {
    return (
        <BradConsumer>
          {value => <span>{value.x}</span>}
        </BradConsumer>
    )
  }
}
```
With code splitting and module state, if the module is entirely self contained within a chunk does it solve the problem you were outlining @timneutkens -- i can confirm that the only time the state for the context is referenced within the compiled code is in the chunk and not within the page or the app computed entry points. Or is it the case that both times it is imported from the various entry points that the module state is essentially initialized again so that even though the code is split out each entry point is creating its own in-memory allocation for the code contained in the common chunk?

**Comment by brad-decker** [comment]
With the above implementation the server renders "3", on the client there is a flash of "loading" followed by "5".

**Comment by brad-decker** [comment]
@timneutkens is the client side bundle just in .next/bundle versus server side in .next/dist/bundle? I'm trying to see how each of these differ in terms of its implementation of the component tree as it relates to the context (specifically the dynamic import variation I did above) and they seem identical. Its as if when the Consumer is being mounted it doesn't see the Provider in the same tree, versus state being an issue.

**Comment by joeporpeglia** [comment]
> Or is it the case that both times it is imported from the various entry points that the module state is essentially initialized again so that even though the code is split out each entry point is creating its own in-memory allocation for the code contained in the common chunk?

Yeah this is the root problem here. For the server compilation only, both `index.js` and `_app.js` are separately bundled into two commonjs modules that are run in node. This server compilation also uses a [special webpack "externals" condition](https://github.com/zeit/next.js/blob/aec4c00214a82097d07295b00f0af5f4eb24167b/server/build/webpack.js#L64-L96) so that all non-source modules (basically anything from `node_modules` that isn't nextjs) are resolved through regular node `require` calls. This prevents webpack from bundling modules into `index.js` and `_app.js` that already work in a nodejs env. However, your `'components/context'` file does not match that exclude (and shouldn't since it must be transpiled). This causes the module to be included in both of the commonjs modules produced for `index.js` and `_app.js`.

Unfortunately I don't think there's any way to create a commons chunk and have it load properly when targeting node. Maybe another solution could involve bundling `_app.js` and `_document.js` with each page module. I think that would also require some changes to how we [import the app, document, and requested page](https://github.com/zeit/next.js/blob/aec4c00214a82097d07295b00f0af5f4eb24167b/server/render.js#L59-L66).

**Comment by brad-decker** [comment]
If there were a way to build a commons chunk look like in my example? Would it exist in the chunks/ folder inside dist? And how would it differ from the one generated by using dynamic imports

**Comment by joeporpeglia** [comment]
@brad-decker I don't believe there's any way to make commons chunks work when targeting node 😞. I also don't think the dynamic imports you provided in that example would work as long as the server renderer `require`'s `_app.js` and the currently requested page separately. https://github.com/zeit/next.js/blob/aec4c00214a82097d07295b00f0af5f4eb24167b/server/render.js#L59-L66

**Comment by steida** [comment]
Sorry to disturb, but why _app is a thing? I have own app component, and context is preserved correctly. What am I missing?

**Comment by joeporpeglia** [comment]
If you render that custom app component as the root in each of your `pages/` then it'll remount when navigating to another page on the client. This makes it difficult to add things like page transitions or global react state across different pages.

**Comment by mtford90** [comment]
For now i've worked around this issue by avoiding _app.js and wrapping every single page like so:

```js
import * as React from "react";
import urlContext from "./urlContext";

export default (PageComponent) => {
  return class Page extends React.Component {
    public static async getInitialProps(initialProps) {
      const { req } = initialProps
      const url = req && req.url;
      let props = { url };

      const getInitialProps = PageComponent.getInitialProps;

      if (getInitialProps) {
        props = {
          ...props,
          ...getInitialProps(initialProps)
        }
      }

      return props;
    }

    public render() {
      const url = this.props.url;

      return (
        <urlContext.Provider value={url}>
          <PageComponent/>
        </urlContext.Provider>
      );
    }
  }
}
```

**Comment by brad-decker** [comment]
@mtford90 yeah, i created a HoC as well for setting up context, but its not clean. I actually refactored away from a HoC when i upgraded to 6 and saw the potential there for the app to handle that for us.

**Comment by elrumordelaluz** [comment]
[here](https://github.com/elrumordelaluz/with-app-layout-app) is a quick repro with the current behaviour in case could be useful for _debug_/ try stuff

**Comment by RustyDev** [comment]
I haven't had time to try yet, but I saw the new [with-redux](https://github.com/zeit/next.js/blob/canary/examples/with-redux/) example that's using `_app.js` and a [provider context](https://github.com/zeit/next.js/blob/canary/examples/with-redux/lib/with-redux-store.js#L27). Looks promising.

**Comment by Ilaiwi** [comment]
Hi, any updates on this issue?

**Comment by ckeeney** [comment]
Does this mean projects following the examples for [`react-redux`](https://github.com/zeit/next.js/tree/canary/examples/with-redux), etc will have broken SSR?  I am having SSR problems after migrating to wrapping my `_app.js` in a `hoc` rather than all my individual pages.

**Comment by timneutkens** [comment]
@ckeeney nope, since they still use the old React context.

**Comment by steida** [comment]
@joeporpeglia What's wrong with stateless app component, where all state is stored in the session? Isn't it a better approach?

**Comment by tz5514** [comment]
Will this issue be resolved if we upgrade webpack to 4.0 in next.js?

**Comment by timneutkens** [comment]
It won't be solved by upgrading webpack.

**Comment by ex3ndr** [comment]
Is there a way to manually put common code to one chunk? It seems that this will solve this problem.

**Comment by filidorwiese** [comment]
Any update? It would be nice be able to use the new Context API consistently on both the server and the client

**Comment by timneutkens** [comment]
No update. I've been working on upgrading to webpack 4.

**Comment by gaearon** [comment]
>>Does this mean projects following the examples for react-redux, etc will have broken SSR? I am having SSR problems after migrating to wrapping my _app.js in a hoc rather than all my individual pages.
>@ckeeney nope, since they still use the old React context.

I wanted to mention this will become more important as heavily used libraries will soon start publishing versions that use the new context API instead.

**Comment by timneutkens** [comment]
We'll be tackling this after webpack 4 lands, as it'll make it easier to implement a fix for this issue.

**Comment by vladanpaunovic** [comment]
@timneutkens when will this happen (approximately)? 😄

**Comment by timneutkens** [comment]
Progress on the webpack 4 branch can be followed here: #4639, there’s no way to give a timeline for this specific issue as this is a bug that has to be investigated to come up with a fix. It has my priority after #4639 lands.

**Comment by timneutkens** [comment]
Basically the issue was exactly what my explanation said here:
https://github.com/zeit/next.js/issues/4194#issuecomment-387443497

This solves it: https://github.com/zeit/next.js/pull/4639/commits/00f62f55907d9d3b47104abe4db04d261d5beabe

**Comment by martinjlowm** [comment]
I'm not entirely sure this is fixed, the shared module part, yes, but not the context behavior.

I just upgraded to `6.1.1-canary.3` (which includes #4639) and it still appears that a provider can fire before a consumer is rendered, resulting in the consumer's function to never render its children.

Compared to the browser, `reactContext._currentValue` appears to remain equal to `reactContext._defaultValue` until the consumer is initially rendered.

When Next.js renders, `reactContext._currentValue` is updated before the consumer element is rendered.

I have a simple case as the follows (yet I haven't tested exactly this example):
```
                                     <- 1.
<reactContext.Provider value="Test">
                                     <- 2.
  <reactContext.Consumer>
    {value => value}
  </reactContext.Consumer>
</reactContext.Provider>
```

In the browser, `_currentValue` is equal to `_defaultValue` in 1. and 2.

When Next.js renders, `_currentValue` is equal to `_defaultValue` in 1. and equal to `"Test"` in 2.

**Comment by gaearon** [comment]
@timneutkens Thank you for jumping on this so quickly!

**Comment by timneutkens** [comment]
@gaearon happy to, was an interesting one to solve 😄

**Comment by jclem** [comment]
Is it possible to solve this problem when using TypeScript? It appears to still be affecting the `with-typescript` example, as in https://github.com/jclem/with-typescript-app-context (notice the difference in logged messages in server vs client).

**Comment by fi5u** [comment]
@timneutkens With this fix, is `Context.Provider` only permitted to be used from __app.js_? For me, it works when I set the provider in _app.js, but `Context.Consumer` does not get the values when `Context.Provider` is defined in a file in _pages_. Is this by design? Ideally, I'd like to set `Context.Provider` at the page, not app, level, then use `Context.Consumer` in child components.

Next version: 7.0.0-canary.20

**Comment by thomhos** [comment]
Just chiming in here. I'm currently experiencing this issue on `next 6.1.2` using `zeit/next-typescript`.

I've created a file which creates a context:
```
export interface TranslationContextProperties {
  language: string;
  translations: TranslationMap;
}

export const TranslationContext = React.createContext<TranslationContextProperties>({
  language: 'en',
  translations: defaultTranslations,
});
```

I then import it in `_app.tsx` and assign a value to `TranslationContext.Provider`.

If I then use a `TranslationContext.Consumer` in the same file it works both client and server side.

However, if I import the context and use the Consumer in the index for instance, it only works client side.

Now I'm wondering, should this be fixed on 6.1.2 or only in 7?

**Comment by timneutkens** [comment]
Only in 7

**Comment by dfoverdx** [comment]
I'm running into this same issue with Next 7.0.2.  I found that if I wrap the context into its own component which won't update its state if the value passed as a property is undefined, that will work between pre-fetched page changes, however that has its own issue of propagating values (for some reason, the inner consumer stays one value behind).

**Comment by ssylvia** [comment]
@timneutkens Also ran into this issue with Next 7.0.2.

**Comment by iMerica** [comment]
This issue persist for me in Next 8.1.0. 

I'm following the official "with redux" example [here](https://github.com/zeit/next.js/tree/master/examples/with-redux) and I continue to see

```
[ wait ]  compiling ...
[ error ] ./pages/_app.js
Module not found: Can't resolve 'react-redux' in '/app/pages'
```
My package.json includes
```json
  "dependencies": {
    "@zeit/next-sass": "^1.0.1",
    "@zeit/next-typescript": "^1.1.1",
    "next": "^8.1.0",
    "node-sass": "^4.12.0",
    "react": "^16.8.6",
    "react-dom": "^16.8.6",
    "react-redux": "^7.0.3",
    "redux": "^4.0.1"
  }
```


This only happens after adding `_app.js`. Regular `./pages/*.js` pages worked fine.

**Comment by dandrei** [comment]
Same problem when using TypeScript.

**Comment by Herlevsen** [comment]
I can also confirm i am having problems with context api in SSR on 8.1.0 and i had to downgrade again.

**Comment by medmin** [comment]
Next 9.0.5, 

I am still NOT  be able to migrate from React SPA to Nextjs, with the React Context API.

And also, the error message is misleading.

I have the error message like this: 

`Uncaught Error: Invariant failed: You should not use <withRouter(BaseContainer) /> outside a <Router>`. 

It took me a week to figure it out that it's NOT  the router. It's the Context!!!!

**Comment by timneutkens** [comment]
@medmin please read this thread: https://twitter.com/timneutkens/status/1154351052973625346

This bug does not exist anymore so it must be something in your code, we have tests verifying it works correctly.

**Comment by sytona** [comment]
What wasn't obvious, was that the implementation apparently defaults to a JavaScript version which would never render its children. I'd expect it to default to CSS ..

**Comment by timneutkens** [comment]
@sytona I'm not sure what you're referring to.

**Comment by sytona** [comment]
Will pack every page into it's own bundle (entrypoint), which ends up in .next/dist. This bundle holds everything, including all imported components etc.

Say you have 2 pages that both import Header it will end up in both page 1 and page 2's bundle.

Because of this module state / singletons don't work, since if you server render page 1, and it holds a module state inside the component, for example,,??

**Comment by timneutkens** [comment]
That's not true because webpack has it's own implementation of the module singleton system. This implementation does have the issue that the modules object is per-entrypoint, however I wrote a webpack plugin that changes it to share the modules object between entrypoints that solves this issue.

**Comment by Ahmdrza** [comment]
Getting same issue in next 9.2.1. Context API only works if consumer is placed in _app.tsx. It does't work when placed in pages.

**Comment by timneutkens** [comment]
> Getting same issue in next 9.2.1. Context API only works if consumer is placed in _app.tsx. It does't work when placed in pages.

This comment is not actionable and on a closed issue from almost 2 years ago.

**Comment by matomesc** [comment]
@timneutkens I'm using `next@9.3.2` and have the same issue with a custom app with `getInitialProps`. It looks like only non-page components directly inside `App.render()` have access to the correct context. When accessing context inside a page, the default context value is returned.

**Comment by platocrat** [comment]
> For now i've worked around this issue by avoiding _app.js and wrapping every single page like so:
> 
> ```js
> // lib/page.js
> import * as React from "react";
> import urlContext from "./urlContext";
> 
> export default (PageComponent) => {
>   return class Page extends React.Component {
>     public static async getInitialProps(initialProps) {
>       const { req } = initialProps
>       const url = req && req.url;
>       let props = { url };
> 
>       const getInitialProps = PageComponent.getInitialProps;
> 
>       if (getInitialProps) {
>         props = {
>           ...props,
>           ...getInitialProps(initialProps)
>         }
>       }
> 
>       return props;
>     }
> 
>     public render() {
>       const url = this.props.url;
> 
>       return (
>         <urlContext.Provider value={url}>
>           <PageComponent/>
>         </urlContext.Provider>
>       );
>     }
>   }
> }
> 
> // pages/index.js
> import page from 'lib/page.js'
> 
> export default page(() => {
>     return (
>         <div>A page</div>
>     )
> })
> ```

What does the `urlContext.js` file contain?

**Comment by hazae41** [comment]
It's still happening in Next.js 11...

**Comment by marktellez** [comment]
I can confirm I am seeing this exact problem in both v10 and v11.

app.jsx wraps the pages with their contexts:

```
<UserProvider>
      <SessionProvider>
        <CartProvider>
          <Component {...pageProps} />
        </CartProvider>
      </SessionProvider>
    </UserProvider>
```

The ages get their contexts loaded just fine via client rendering/navigation with useRouter.

If there is a SSRedirect or a user enters a url directly, the state that is supposed to be available from the context is "NULL".

**Comment by timneutkens** [comment]
@hazae41 @markhaslam please create a new issue with a complete reproduction so that it can be investigated. The initially reported issue has tests in the Next.js test suite.
